### PR TITLE
Convert dials.generate_mask to functional interface

### PR DESCRIPTION
**Pull request by benjaminhwilliams**
Following our discussion around #691, I thought I'd start a few bits of functional interface conversion, specifically the bits that affect i19.screen.  To start tamely, I've done `dials.generate_mask`.  I'd quite like to have it reviewed before I do any of the others, to check that I'm not doing anything controversial.

One particular question — I've converted some `Sorry`s to `ValueError`s, but should I be using `sys.exit()` in this context?

As far as possible, the changes follow boilerplate agreed in [the wiki](../wiki/Boilerplate-(or-the-standard-framework-of-a-DIALS-program)).

### CODE REVIEW

**Comment by jbeilstenedmands** [comment]
Regarding sys.exit() vs ValueError I think you have it correct at the moment. I believe sys.exit will not give a traceback, which is nice for catching simple user errors at the highest level, but maybe the tracebacks are more helpful deeper within the code to aide debugging.

**Comment by benjaminhwilliams** [comment]
Indeed.  They were `Sorry`s before though, which suggested that suppressed traceback was the intention.  But I don't know if it was a good intention...  Happy to leave this to @Anthchirp's opinion.

File: command_line/generate_mask.py [hunk 1]
```diff
@@ -32,108 +30,157 @@
...
+    """
+    imagesets = experiments.imagesets()
+    if len(imagesets) > 1:
+        # Check beams (for resolution) and detectors are equivalent in each case
+        # otherwise the mask may not be appropriate across all imagesets
+        detectors = experiments.detectors()
+        beams = experiments.beams()
+        for d in detectors[1:]:
+            if not d.is_similar_to(detectors[0]):
+                raise ValueError(
```
Comment (rjgildea) [inline command_line/generate_mask.py]: I would argue that the user shouldn't be exposed to a full traceback here.

File: command_line/generate_mask.py [hunk 2]
```diff
@@ -32,108 +30,157 @@
...
+                    "Multiple imagesets are present, but their detector models differ."
+                )
+        for b in beams[1:]:
+            if not b.is_similar_to(beams[0]):
+                raise ValueError(
+                    "Multiple imagesets are present, but their beam models differ."
+                )
```
Comment (rjgildea) [inline command_line/generate_mask.py]: As above, I would argue that the user shouldn't be exposed to a full traceback here.

File: command_line/generate_mask.py [hunk 3]
```diff
@@ -32,108 +30,157 @@
...
-        if params.output.experiments is not None:
-            for imageset in imagesets:
-                imageset.external_lookup.mask.data = ImageBool(mask)
-                imageset.external_lookup.mask.filename = params.output.mask
-            from dxtbx.model.experiment_list import ExperimentListDumper
-
-            print("Saving experiments to {0}".format(params.output.experiments))
-            dump = ExperimentListDumper(experiments)
-            dump.as_file(params.output.experiments)
+def script(experiments, params):
```
Comment (ndevenish) [inline command_line/generate_mask.py]: Did we specifically agree that the function should be called `script`? Surely something more descriptive like `generate_mask` would be a lot more intuitive e.g.
```python
from dials.command_line.generate_mask import generate_mask

# later...
generate_mask(exp, params)
"""

Comment (rjgildea) [reply command_line/generate_mask.py]: I agree - don't see the point in having a meaningless class/function name `script` proliferated across the source code.

File: command_line/generate_mask.py (hunk 1 above)
Comment (benjaminhwilliams) [reply command_line/generate_mask.py]: OK, I'll use sys.exit instead.

```suggestion
                sys.exit(
```

File: command_line/generate_mask.py (hunk 3 above)
Comment (benjaminhwilliams) [reply command_line/generate_mask.py]: Yep, sorry, I was cargo culting.  I'll give it a proper descriptive name.

File: command_line/generate_mask.py (hunk 2 above)
Comment (benjaminhwilliams) [reply command_line/generate_mask.py]: Agreed as above.

File: command_line/generate_mask.py [hunk 4]
```diff
@@ -32,108 +30,157 @@
...
+            containing only one :type:`dxtbx_model_ext.Experiment` object.
+        params (:type:`libtbx.phil.scope_extract`): Masking parameters, having the
+            structure defined in :data:`dials.util.masking.phil_scope`.
+
+    Returns:
+        Tuple[:rtype:`scitbx_array_family_flex_ext.bool`]: A tuple containing the
+            generated pixel masks.
+        :rtype:`dxtbx_model_ext.ExperimentList` (optional): A copy of
+            :param:`experiments` with the masks applied.  Only returned if
+            :attr:`params.output.experiments` is set.
```
Comment (ndevenish) [inline command_line/generate_mask.py]: [**tl;dr** - type is probably wrong for return mask, also return type in general]

Couple of Docstring notes; not really criticism as points not raised before:
- I don't think you need to explicitly link the types as types in the type field. At least, it should know what these are types already? Even if it doesn't, it should just be a matter of parser intelligence. Made redundant by,
- Probably a bigger discussion but I've started moving even more to Python3 style type annotations. The main checking tool is [mypy](http://mypy-lang.org/), which has a standard for [python2](https://mypy.readthedocs.io/en/stable/cheat_sheet.html) _**But** also_ a way to automatically convert the python2-style annotations to real [PEP-484](https://www.python.org/dev/peps/pep-0484/) annotations once the codebase is ready for python3-only, so it's a bit more future-compatible.
- On this note, there are definitely plugins for sphinx/napoleon that will read the type annotations and extract them automatically so that you don't need to define them in the actual docstring text. This means that they only ever need to be written once, and in a static-analysis-verifiable way.

Specific to this code point:
- Assuming that this really does return a tuple of multiple masks rather than a tuple of a single mask, you probably mean `Tuple[scitbx.array_family.flex.bool, ...]` (and actual import path should be fine)
- The return of a type that could be multiple things is
  - In sphinx notation `A or B`
  - In PEP-484 notation: `Union[A, B]`

So, the return type is:
```
Tuple[scitbx.array_family.flex.bool,...] or Tuple[Tuple[scitbx.array_family.flex.bool,...], dxtbx.model.ExperimentList]
or
Union[Tuple[Tuple[scitbx.array_family.flex.bool,...], dxtbx.model.ExperimentList], Tuple[scitbx.array_family.flex.bool,...]]
```
And the full type-annotation type for the function, mypy2-style is:
```python
# type: (dxtbx.model.ExperimentList, libtbx.phil.scope_extract) -> Union[Tuple[Tuple[scitbx.array_family.flex.bool,...], dxtbx.model.ExperimentList], Tuple[scitbx.array_family.flex.bool,...]]
```
or

```python
# mypy python 2-style type annotation:
def generate_mask(
    experiments,  # type: dxtbx.model.ExperimentList
    params  # type: libtbx.phil.scope_extract
):
    # type: (...) -> Union[Tuple[Tuple[scitbx.array_family.flex.bool,...], dxtbx.model.ExperimentList], Tuple[scitbx.array_family.flex.bool,...]]
    """
    Generate a pixel mask for each image in an experiment.

    ...
    """
```
which becomes, in python 3:
```python
def generate_mask(
    experiments: dxtbx.model.ExperimentList, params: libtbx.phil.scope_extract
) -> Union[
    Tuple[Tuple[scitbx.array_family.flex.bool, ...], dxtbx.model.ExperimentList],
    Tuple[scitbx.array_family.flex.bool, ...],
]:
    """
    Generate a pixel mask for each image in an experiment.

    ...
    """
```

This looks... not good, but can generally be simplified by scoped imports and custom return types e.g.
```python
from dxtbx.model import ExperimentList
Masks = Tuple[scitbx.array_family.flex.bool, ...]

def generate_mask(
    experiments: ExperimentList, params: libtbx.phil.scope_extract
) -> Union[Masks, Tuple[Masks, ExperimentList]]:
    """Docstring"""
```

This can obviously be reduced further if the return type is simplified to something like always returning a tuple (which is easier to use also) e.g.`Tuple[Masks, Optional[ExperimentList]]`

Comment (ndevenish) [reply command_line/generate_mask.py]: The other point about not having lots of reST markup in the type field is that google/numpy style docstrings are specifically designed to be easy to read inline in code, which obviously the markup somewhat obfuscates

Comment (benjaminhwilliams) [reply command_line/generate_mask.py]: Thanks, I got a bit confused when reading the various different specifications, so that's very helpful.  I'll go through again and strip out some chaff.

The returned object is indeed, in general, a tuple of multiple mask objects.  As to always returning a tuple with an optional second value, you have a good point.  I think I tried to over-engineer it a bit because I once got caught out by https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.polyfit.html#numpy-polyfit

File: command_line/generate_mask.py (hunk 2 above)
Comment (ndevenish) [reply command_line/generate_mask.py]: I know you'll hate me but I don't think sys.exit is the right thing either. It shows the message, but will also cause an unexpected exit if being used functionally.

Possible alternative: Raise `Sorry` which is an exception but also doesn't stacktrace, or throw something custom and catch it in the `run` function (don't want to hide any exceptions coming from underneath the `generate_mask` function).

File: command_line/generate_mask.py (hunk 4 above)
Comment (ndevenish) [reply command_line/generate_mask.py]: Ah, maybe the numpy style that explicitly means multiple/tuple? From the linked document I can't tell if it's one or the other or a tuple of both or what.

The typing module (in 3, but backported) makes stuff nicer, because all these type annotations may look slightly verbose but are completely legal python expressions so statically verifiable, which is why you can do stuff like `Masks = Tuple[flex.bool,...]` and have it work.

Comment (ndevenish) [reply command_line/generate_mask.py]: But as an ambition, the actual markup the typing should make it unambiguous, and then whatever documentation generator can render it how it likes.

**Comment by ndevenish** [comment]
from Traceback Terror to Exit Exceptionalism?

File: command_line/generate_mask.py (hunk 4 above)
Comment (benjaminhwilliams) [reply command_line/generate_mask.py]: Hmm... do we want to specify `IO[str]` or similar for functions that write to disk?

Comment (benjaminhwilliams) [reply command_line/generate_mask.py]: As in https://mypy.readthedocs.io/en/latest/cheat_sheet_py3.html#miscellaneous

Comment (ndevenish) [reply command_line/generate_mask.py]: Not sure if I've misunderstood the question, but that's still for taking parameters/returning values - just if it's a file-like object (e.g. something that has .write, .read, .seek, .flush maybe etc). So, for example, the `show` methods as discussed in https://github.com/cctbx/cctbx_project/issues/275 often take a `log` parameter that it writes to, and they would be of type `IO`. The PEP seems to specify it for *just* things that write to disk but people seem to use it more generally.

Now, if you want to actually encode the behaviour of a function's side effects into the type system, then you start getting into ML style languages and monads and what have you......

**Comment by benjaminhwilliams** [comment]
If @ndevenish is happy with typing & docstrings, I have one question for @Anthchirp — did we arrive at a conclusion as to whether we should be using `libtbx.phil` or `iotbx.phil`?  I haven't delved.

File: command_line/generate_mask.py (hunk 4 above)
Comment (benjaminhwilliams) [reply command_line/generate_mask.py]: Ah, yep.  I skim-read and didn't correctly comprehend.  Ignore the question!

**Comment by rjgildea** [comment]
If you need to parse e.g. space groups or unit cells from phil then you’ll need to use iotbx.phil, otherwise libtbx.phil is fine.

**Comment by ndevenish** [comment]
Made a few further tweaks to this in 059a88bd09a77b1a8abf94edfa81cbc30473e8eb.

- Moved phil functions back into a scoped import. Reasoning: It's reasonable to expect any coder looking at dials to know what an `ExperimentList` is, but less so `parse` or `scope` in some random context. Having it prefixed as `phil.parse` makes it obvious what is being called.
- The `typing` module is an optional backport in python2, so we can't necessarily assume it's there as I'm not sure what pulls it in. Protect this with a try block.
- Couple of type and docstring description tweaks. E.g. `list` implies an actual list object, while `List` technically is something very list-like (though the [docs](https://docs.python.org/3/library/typing.html#typing.List) ties them pretty closely). In any case, `List` seems to be preferred in these contexts. Also, lower-case looks weird in types 😜

**Comment by ndevenish** [comment]
re: the sys.exist()/exception thing - also briefly mentioned with @benjaminhwilliams but not explored, technically this only accepts `ExperimentLists` with one experiment. It even checks in the `run` function.

I think this means that it can literally only ever have one imageset/detector/etc, meaning the checks can be removed? It might be worth adding in an assert to the functional function, or, perhaps, switching it to only take a single `Experiment`?

**Comment by benjaminhwilliams** [comment]
OK, you're correct but this raises a broader point of multi-sweep data collections being effectively treated as something of a poor relation, which causes headaches when working with typical small molecule data.  I would prefer to have `generate_mask` ingest an `ExperimentList` and for the mask generation code to be re-written to accommodate multiple imagesets.  That's an issue (#604) outside the scope of this PR though.

**Comment by ndevenish** [comment]
Okay so realised with @benjaminhwilliams that it's not multiple experiments per-se that are the problem but multiple imagesets, so in 25a8faff10026aa42a9df0774a1c7a0f04dbce49 changed docs to reflect that, and check explicitly with an assert. This means that the checks for multiple imageset features were never going to be hit, so removed. 

In the future any better solution will have to work round these problems but with different code anyway so 🤷‍♂️

**Comment by benjaminhwilliams** [comment]
Hmm, @ndevenish I think 25a8faf breaks @dagewa's original fix to #604.  Multiple imagesets can share a single beam and detector, in which case the mask generated is the same for all imagesets.

Proposed solution, either:
 - Revert 25a8faf, merge PR, fix #604 properly later; or
 - Make merging this PR dependent on fixing #604 in this branch.

Since, as I've already mentioned, #604 is somewhat off topic, my preference is for the former.

**Comment by ndevenish** [comment]
Okay, I think I see what has happened. My commit *does* break @dagewa's commit, but only if it wasn't already broken, which it was.

I was basing my logic on the "only one experiment" condition.

Except this condition was only written in https://github.com/dials/dials/commit/747e2410b272d6576ff87086893bd237f26de298#diff-081736f62f9fa5765da636acc43d54c7L83 a couple of months ago when @jmp1985 changed datablocks to experiment lists. But I think this was wrong, because whereas `ExperimentList` cannot contain more than one imageset, I believe `Datablock` could, so changing the check on len() == 1 wasn't correct.

So, yes, there needs to be some reversion but also some rewriting.

**Comment by benjaminhwilliams** [comment]
Ah, good detective work.  OK, I'll do the legwork.

For the record, an `Experiment` can contain only one imageset, but an `ExperimentList` can contain more than one imageset because it can contain more than one `Experiment`.

**Comment by ndevenish** [comment]
Ah yes, obviously, thanks.

File: command_line/generate_mask.py [hunk 5]
```diff
@@ -85,55 +86,70 @@
...
     """
     imagesets = experiments.imagesets()
-    assert len(imagesets) == 1, "Only single imagesets supported"
-
-    imageset = imagesets[0]
+    masks = []
+
+    # Specify the formatting of the output filenames
+    num_imagesets = len(imagesets)
+    pad = len("{:d}".format(num_imagesets))
```
Comment (ndevenish) [inline command_line/generate_mask.py]: `pad = len(str(num_imagesets))` a bit clearer? I don't think the `:d` actually buys you anything here?

Comment (ndevenish) [reply command_line/generate_mask.py]: Also this is a more picky comment but I'm not sure that `num_imagesets = len(imagesets)` is saving much - only one less character for the extra indirection.

Comment (benjaminhwilliams) [reply command_line/generate_mask.py]: Good point, thanks.  I'll push a change when you're done reviewing.

File: command_line/generate_mask.py [hunk 6]
```diff
@@ -85,55 +86,70 @@
...
+    # If there is more than one imageset, append a number to each output mask filename
+    filenames = [
+        "_{:0{:d}d}".format(i + 1, pad).join(os.path.splitext(params.output.mask))
+        if num_imagesets > 1
+        else params.output.mask
+        for i in range(num_imagesets)
+    ]
```
Comment (ndevenish) [inline command_line/generate_mask.py]: Had to do some mental contortions for this loop - both the `join(os.path.splitext(...` is a bit of an unusual way to do that (though, clever!). Also found the embedded ternary that only happens if there is only one item a little hard to follow.

Maybe something like:
```python
    # Create output mask filenames
    if num_imagesets == 1:
        filenames = [params.output.mask]
    else:
        # If there is more than one imageset, append a number to each output mask filename
        name, ext = os.path.splitext(params.output.mask)
        pad = len(str(num_imagesets))
        filenames = [
            "{name}_{num:0{pad}}{ext}".format(name=name, num=i + 1, pad=pad, ext=ext)
            for i in range(num_imagesets)
        ]
```
or, if you feel the format expansion doesn't need the naming:
```python
        filenames = [
            "{}_{:0{}}{}".format(name, i + 1, pad, ext) for i in range(num_imagesets)
        ]
```
There's possibly an argument that this could be moved off into a small internal function but I'm ambivalent about that in this case.

One other minor point: the `d` typing in format is redundant - if you are passing in a standard integer then it's the default, and it won't e.g. cause a stringified integer or float to convert. I guess there's a question of why it's in the standard if you never need it, but I'd imagine

1. printf-style has it (as printf *needs* the type in the string to correctly decode) and kept for sanity
2. You may want e.g. a custom class that accepts integer-style formatting but defaults to hex or something

As an aside, *knowing* about f-strings really makes me want to write them:
```python
filenames = [f"{name}_{i+1:0{pad}}{ext}" for i in range(num_imagesets)]
```

File: test/command_line/test_generate_mask.py [hunk 7]
```diff
@@ -33,7 +49,6 @@ def test_generate_mask_with_untrusted_rectangle(input_filename):
     assert result["stderr"] == ""
     assert os.path.exists("mask2.pickle")
     assert os.path.exists("masked_experiments.json")
-    from dxtbx.serialize import load
```
Comment (ndevenish) [inline test/command_line/test_generate_mask.py]: Like +++

**Comment by benjaminhwilliams** [comment]
I think this has probably been reviewed enough ;-p

Merging with prejudice.
